### ISSUE TITLE
Replace `Transform.__add__` with `Transform.__matmul__`?

### ISSUE THREAD

**Issue by brunobeltran** [matplotlib/matplotlib#17230]
### Bug report

**Bug summary**

While working on fixing the transform pipeline for image rendering, it struck me that the operator "+" probably sends the wrong impression to users in terms of expected properties. The Euclidean group is barely associative, and definitely not commutative. 

Also, since a `Transform` is something that's *applied* (like an operator), my expectation as I'm using composite transforms is that they will be applied rightmost-first. In other words, I keep expecting

```python
>>> (A + B).transform(X) == A.transform(B.transform(X))  # but it's currently False
>>> (A + B).transform(X) == B.transform(A.transform(X))  # how it currently works.
```

Now that Python has the "@" operator, this feels like a more natural choice for "combining" `Transforms`. People don't generally expect `A@B == B@A`, while they do expect `A + B == B + A`, which is definitely not true for Transforms. 

The implementation would just be something like (modulo deprecation code)

```python
def __matmul__(self, other):
    return other + self
```

And we would get 
```python
>>> (A @ B).transform(X) == A.transform(B.transform(X))  # returns True
```

**Release strategy** 

Fixing all our internal uses is (probably) only one PR's worth of work, since (relative to how much code would churn) it would be easy to verify/review. 

However, to me this feels like an integral enough part of the Transform API that it might make sense to leave in "+" until 4.x, and maybe just make sure we only advertise "@" in e.g. the tutorials?

In summary:

**Pros**
1. "@" implies lack of commutivity, whereas "+" implied commutivity
2. "@" implies order of application (rightmost-first), whereas "+" is ambigious

**Cons**
1. Pretty core part of `Transform` API
2. Lots of code-churn in `Transform`-related internals.

**Comment by brunobeltran** [comment]
Oh, forgot the main reason why I started being bothered by this. 

A lot of places in the code use "+=" to combine transforms, such as in `_make_image`, where we do 

```python
            t += Affine2D().scale(1.0 + extra_width, 1.0 + extra_height)
```

First off, I can't find where we overwrite `__iadd__`, so I'm a bit confused as to how that even works..., but that aside, whether this gets added on the right or left *matters*. So having it be possibly ambiguous with "+=" makes life harder for someone trying to read/understand this code.

**Comment by anntzer** [comment]
This was *just* discussed in https://github.com/matplotlib/matplotlib/pull/17206#issuecomment-617902548 / https://github.com/matplotlib/matplotlib/issues/17217 / https://github.com/matplotlib/matplotlib/pull/17218.  I personally think `+`/`-` actually work better here (well, in particular there's no `-` for `@` (aka no left-divide) and that's a real problem).

**Comment by brunobeltran** [comment]
Oh sorry, github, didn't recommed those to me, I'll close this.

**Comment by brunobeltran** [comment]
Seems like the main con is actually losing the ability to cleverly make

```python
A @ A.inverted() == IdentityTransform()
```

return `True`.

So in order to get feature parity, we'd need `.inverted` to return an `InvertedTransform` of some sort so that `CompositeGenericTransform` can check if it cancels exactly when evaluating "@". I'll reopen this if people think that's worth it....but it seems like too much work for too little benefit to me now.

P.S.
Seems like an additional complaint is also that `B**-1 @ A` and `(1/B) @ A` aren't as "pretty" as "A - B". I disagree, since we don't typically have long chains of transforms to make it unwieldy to just type `B.inverted() @ A`. And if we do, I think that's honestly more readable than `B - A`. The latter is now well documented (thanks @anntzer!) but to be fair, the former doesn't even need documentation. It just works the way you think it does.

**Comment by anntzer** [comment]
Having a real InvertedTransform() was proposed at least once (by @ImportanceOfBeingErnest iirc? you'll have to look it up) to get a non-frozen inverse() (i.e. one that keeps track of changes in the original thing); I think it would be usefule, and I guess it would be a prerequisite for a switch to matmul to happen (even though right now I still don't like it).

**Comment by brunobeltran** [comment]
Looks like it was proposed in #10741.
